common/nglinks.py

```diff
@@ -0,0 +1,116 @@
+"""Neuroglancer link functions."""
+import json
+import urllib
+from typing import Optional, Union
+from collections import OrderedDict
+from dataclasses import dataclass
+from enum import Enum
+
+
+class LayerType(Enum):
+    Image = "image"
+    Segmentation = "segmentation"
+
+
+@dataclass
+class Layer:
+    cloudpath: str
+    type: LayerType
+
+
+class ImageLayer(Layer):
+
+    def __init__(self, cloudpath: str):
+        self.cloudpath = cloudpath
+        self.type = LayerType.Image
+
+
+class SegLayer(Layer):
+
+    def __init__(self, cloudpath: str):
+        self.cloudpath = cloudpath
+        self.type = LayerType.Segmentation
+
+
+def generate_ng_payload(
+    layers: dict[str, Layer], center: Optional[tuple[int, int, int]] = None
+) -> str:
+    """Makes neuroglancer link payloads from mappings between layer names and paths.
+
+    Makes the viewer resolution equal to the "img" layer if it exists in layers.
+    Otherwise, it takes the first it can find.
+    """
+    assert len(layers) > 0, "empty layer paths"
+
+    ngl_layers = OrderedDict()
+    layers = layers.copy()  # for in-place modifications later
+
+    # default to first resolution we can find - update if we find an image
+    ng_resolution = dataset_resolution(list(layers.values())[0])
+
+    # Trying to find an image layer
+    base_img = layers.pop("img", None)
+    if base_img:
+        ngl_layers["img"] = {
+            "source": fullpath(base_img),
+            "type": base_img.type.value,
+        }
+        ng_resolution = dataset_resolution(base_img)
+
+    for key, layer in layers.items():
+        ngl_layers[key] = {
+            "source": fullpath(layer),
+            "type": layer.type.value,
+        }
+
+    navigation = {"pose": {"position": {"voxelSize": ng_resolution}}, "zoomFactor": 4}
```

Comment:
I wonder if there is a less arbitrary way to choose zoomFactor, 4 seems be fine for most cases though.

Reply:
One way this might work is to use the smallest volume size, but downloading all of the CloudVolume bounds seems slow. I left this out for now.